### PR TITLE
Package opine.0.1.3

### PR DESCRIPTION
**Pull request by ArulselvanMadhavan**
### `opine.0.1.3`
Python AST unparse implementation in OCaml
Opine is an OCaml library that unparses the python AST back to python code.

---
* Homepage: https://github.com/ArulselvanMadhavan/opine
* Source repo: git+https://github.com/ArulselvanMadhavan/opine.git
* Bug tracker: https://github.com/ArulselvanMadhavan/opine/issues

---
:camel: Pull-request generated by opam-publish v2.2.0

### CODE REVIEW

**Comment by ArulselvanMadhavan** [comment]
I think I need help understanding why the CI build is failing

**Comment by ArulselvanMadhavan** [comment]
This is my first time trying to publish a package and I am not sure I'm doing it right

https://opam.ocaml.org/doc/Packaging.html

**Comment by raphael-proust** [comment]
@ArulselvanMadhavan releasing a package is not the easiest thing, don't worry!

I'll have a look at the CI. Even before looking, I can already say that there are some acceptable failures (i.e., some failures that we can still merge with).

**Comment by raphael-proust** [comment]
Ok so here are the failures:

- Some linking failures in the experimental macos+ocaml5 runner. This is likely not directly related to your project. I'll check more later but it seems to be related to `base`.

- Some lower-bound checks where some record-field is not bound. Here's how you can find out what is wrong in this case.

1. The error is `Unbound record field end_line` in a record with a type defined in `Parser.Error`.
2. Looking at your code, the `Parser.Error` module is an alias for pyre-ast's `TaglessFinal.Parsing.ParseError` module, so the record type is actually defined there.
3. The [`git blame` information on the file that defines this module](https://github.com/grievejia/pyre-ast/blame/trunk/lib/taglessFinal/parsing.ml#L43) shows a change in the definition of the record.
4. That change is [a commit titled “Parser error now contains a location range instead of just the starting”](https://github.com/grievejia/pyre-ast/commit/50deaf8a8773f4e06d768f366a3c2b66aa8ddda5) which seems related.
5. On further inspection, it is the commit which adds the field you are using in your code base.
6. That commit is part of the 1.6 release and so you need to add a minimal version constraint of 1.6 to your pyre-ast dependency.

In summary:
- look what the error is
- find out what dependency the error is related
- find out what version of the dependency introduced the feature you are dependent on
- add the corresponding constraint in your opam file

File: packages/opine/opine.0.1.3/opam [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+bug-reports: "https://github.com/ArulselvanMadhavan/opine/issues"
+depends: [
+  "ocaml" {>= "4.14.0"}
+  "dune" {>= "3.6"}
+  "ppx_hash"
+  "ppx_compare"
+  "ppx_sexp_conv"
+  "ppx_make"
+  "ppx_deriving"
+  "pyre-ast"
```
Comment (raphael-proust) [inline packages/opine/opine.0.1.3/opam:21]:
```suggestion
  "pyre-ast" {>= "1.6"}
```

File: packages/opine/opine.0.1.3/opam [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+bug-reports: "https://github.com/ArulselvanMadhavan/opine/issues"
+depends: [
+  "ocaml" {>= "4.14.0"}
+  "dune" {>= "3.6"}
+  "ppx_hash"
+  "ppx_compare"
+  "ppx_sexp_conv"
+  "ppx_make"
+  "ppx_deriving"
+  "pyre-ast" {>= "1.6"}
```
Comment (raphael-proust) [inline packages/opine/opine.0.1.3/opam:21]: oooops I made a mistake:

```suggestion
  "pyre-ast" {>= "0.1.6"}
```

**Comment by ArulselvanMadhavan** [comment]
Thank you @raphael-proust for your helpful comments on how to narrow down the source of the problem!

File: packages/opine/opine.0.1.3/opam [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  "ppx_hash"
+  "ppx_compare"
+  "ppx_sexp_conv"
+  "ppx_make"
+  "ppx_deriving"
+  "pyre-ast" {>= "0.1.6"}
+  "stdio"
+  "base" {>= "0.15.0"}
+  "uutf"
+  "cmdliner"
```
Comment (kit-ty-kate) [inline packages/opine/opine.0.1.3/opam:25]:
```suggestion
  "cmdliner" {>= "1.1.0"}
```
```

#=== ERROR while compiling opine.0.1.3 ========================================#
# context              2.2.0~alpha~dev | linux/x86_64 | ocaml-base-compiler.4.14.1 | pinned(https://github.com/ArulselvanMadhavan/opine/archive/refs/tags/0.1.3.tar.gz)
# path                 ~/.opam/4.14/.opam-switch/build/opine.0.1.3
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p opine -j 127 @install
# exit-code            1
# env-file             ~/.opam/log/opine-7-a135d4.env
# output-file          ~/.opam/log/opine-7-a135d4.out
### output ###
# File "/home/opam/.opam/4.14/lib/ocaml-compiler-libs/bytecomp/ocaml-compiler-libs.bytecomp.dune", line 1, characters 0-0:
# Warning: .dune files are ignored since 2.0. Reinstall the library with dune
# >= 2.0 to get rid of this warning and enable support for the subsystem this
# library provides.
# File "/home/opam/.opam/4.14/lib/ocaml-compiler-libs/common/ocaml-compiler-libs.common.dune", line 1, characters 0-0:
# Warning: .dune files are ignored since 2.0. Reinstall the library with dune
# >= 2.0 to get rid of this warning and enable support for the subsystem this
# library provides.
# File "/home/opam/.opam/4.14/lib/ocaml-compiler-libs/shadow/ocaml-compiler-libs.shadow.dune", line 1, characters 0-0:
# Warning: .dune files are ignored since 2.0. Reinstall the library with dune
# >= 2.0 to get rid of this warning and enable support for the subsystem this
# library provides.
# File "/home/opam/.opam/4.14/lib/ocaml-compiler-libs/toplevel/ocaml-compiler-libs.toplevel.dune", line 1, characters 0-0:
# Warning: .dune files are ignored since 2.0. Reinstall the library with dune
# >= 2.0 to get rid of this warning and enable support for the subsystem this
# library provides.
# File "/home/opam/.opam/4.14/lib/ppx_derivers/ppx_derivers.dune", line 1, characters 0-0:
# Warning: .dune files are ignored since 2.0. Reinstall the library with dune
# >= 2.0 to get rid of this warning and enable support for the subsystem this
# library provides.
# (cd _build/default && /home/opam/.opam/4.14/bin/ocamlopt.opt -w -40 -g -I bin/.main.eobjs/byte -I bin/.main.eobjs/native -I /home/opam/.opam/4.14/lib/base -I /home/opam/.opam/4.14/lib/base/base_internalhash_types -I /home/opam/.opam/4.14/lib/base/caml -I /home/opam/.opam/4.14/lib/base/shadow_stdlib -I /home/opam/.opam/4.14/lib/cmdliner -I /home/opam/.opam/4.14/lib/parsexp -I /home/opam/.opam/4.14/lib/ppx_compare/runtime-lib -I /home/opam/.opam/4.14/lib/ppx_deriving/runtime -I /home/opam/.opam/4.14/lib/ppx_hash/runtime-lib -I /home/opam/.opam/4.14/lib/ppx_sexp_conv/runtime-lib -I /home/opam/.opam/4.14/lib/pyre-ast -I /home/opam/.opam/4.14/lib/pyre-ast/__private__/taglessFinal -I /home/opam/.opam/4.14/lib/result -I /home/opam/.opam/4.14/lib/sexplib -I /home/opam/.opam/4.14/lib/sexplib0 -I /home/opam/.opam/4.14/lib/stdio -I /home/opam/.opam/4.14/lib/uchar -I /home/opam/.opam/4.14/lib/uutf -I lib/.opine.objs/byte -I lib/.opine.objs/native -intf-suffix .ml -no-alias-deps -o bin/.main.eobjs/native/dune__exe__Main.cmx -c -impl bin/main.ml)
# File "bin/main.ml", line 101, characters 13-21:
# 101 |   let info = Cmd.info "expression" ~doc ~man in
#                    ^^^^^^^^
# Error: Unbound module Cmd
```

File: packages/opine/opine.0.1.3/opam [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  "ocaml" {>= "4.14.0"}
+  "dune" {>= "3.6"}
+  "ppx_hash"
+  "ppx_compare"
+  "ppx_sexp_conv"
+  "ppx_make"
+  "ppx_deriving"
+  "pyre-ast" {>= "0.1.6"}
+  "stdio"
+  "base" {>= "0.15.0"}
```
Comment (kit-ty-kate) [inline packages/opine/opine.0.1.3/opam:23]:
```suggestion
  "base" {>= "v0.15.0"}
```

File: packages/opine/opine.0.1.3/opam [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+opam-version: "2.0"
+synopsis: "Python AST unparse implementation in OCaml"
+description:
+  "Opine is an OCaml library that unparses the python AST back to python code."
+maintainer: ["arulselvan1234@gmail.com"]
+authors: ["Arulselvan Madhavan"]
+license: "MIT"
+tags: ["topics" "ocaml" "python" "ast"]
+homepage: "https://github.com/ArulselvanMadhavan/opine"
+doc: "https://url/to/documentation"
```
Comment (kit-ty-kate) [inline packages/opine/opine.0.1.3/opam:11]: If you do not have any pre-generated documentation somewhere the `doc` field should be omited
```suggestion
```

File: packages/opine/opine.0.1.3/opam [hunk 6]
```diff
@@ -0,0 +1,52 @@
+# This file is generated by dune, edit dune-project instead
+opam-version: "2.0"
+synopsis: "Python AST unparse implementation in OCaml"
+description:
+  "Opine is an OCaml library that unparses the python AST back to python code."
```
Comment (kit-ty-kate) [inline packages/opine/opine.0.1.3/opam:5]: The description talks about an OCaml library, however looking at the code, the `opine` library is not installed. Could you add the missing `(public_name opine)` ?

Comment (ArulselvanMadhavan) [reply packages/opine/opine.0.1.3/opam:5]: Yes, I did. Didn't realize that's how libraries were installed via dune/opam. I tested in my local machine by running ```bash opam install . ``` and importing ```Opine``` as a library. So, I'm hoping that this will work!

File: packages/opine/opine.0.1.3/opam [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+bug-reports: "https://github.com/ArulselvanMadhavan/opine/issues"
+depends: [
+  "ocaml" {>= "4.14.0"}
+  "dune" {>= "3.6"}
+  "ppx_hash"
+  "ppx_compare"
+  "ppx_sexp_conv"
+  "ppx_make"
+  "ppx_deriving"
+  "pyre-ast" {>= "0.6.0"}
```
Comment (kit-ty-kate) [inline packages/opine/opine.0.1.3/opam:20]: No version 0.6.0 of pyre-ast exist. Did you mean 0.1.6 or something?

Comment (ArulselvanMadhavan) [reply packages/opine/opine.0.1.3/opam]: I'm sorry about that. Yes, I did intend to write 0.1.6. Fixed it!

**Comment by mseri** [comment]
On macos it fails with
```

#=== ERROR while compiling opine.0.1.3 ========================================#
# context              2.1.4 | macos/x86_64 | ocaml-base-compiler.4.14.1 | pinned(https://github.com/ArulselvanMadhavan/opine/archive/refs/tags/0.1.3.tar.gz)
# path                 ~/.opam/4.14.1/.opam-switch/build/opine.0.1.3
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p opine -j 11 @install
# exit-code            1
# env-file             ~/.opam/log/opine-98485-ee9999.env
# output-file          ~/.opam/log/opine-98485-ee9999.out
### output ###
# File "examples/dune", line 3, characters 7-11:
# 3 |  (name main)
#            ^^^^
# (cd _build/default && /Users/mac1000/.opam/4.14.1/bin/ocamlopt.opt -w -40 -g -o examples/main.exe /Users/mac1000/.opam/4.14.1/lib/base/base_internalhash_types/base_internalhash_types.cmxa -I /Users/mac1000/.opam/4.14.1/lib/base/base_internalhash_types /Users/mac1000/.opam/4.14.1/lib/base/caml/caml.cmxa /Users/mac1000/.opam/4.14.1/lib/sexplib0/sexplib0.cmxa /Users/mac1000/.opam/4.14.1/lib/base/shadow_stdlib/shadow_stdlib.cmxa /Users/mac1000/.opam/4.14.1/lib/base/base.cmxa -I /Users/mac1000/.opam/4.14.1/lib/base /Users/mac1000/.opam/4.14.1/lib/pyre-ast/__private__/taglessFinal/taglessFinal.cmxa -I /Users/mac1000/.opam/4.14.1/lib/pyre-ast/__private__/taglessFinal /Users/mac1000/.opam/4.14.1/lib/ppx_sexp_conv/runtime-lib/ppx_sexp_conv_lib.cmxa /Users/mac1000/.opam/4.14.1/lib/ppx_compare/runtime-lib/ppx_compare_lib.cmxa /Users/mac1000/.opam/4.14.1/lib/ppx_hash/runtime-lib/ppx_hash_lib.cmxa /Users/mac1000/.opam/4.14.1/lib/pyre-ast/pyreAst.cmxa /Users/mac1000/.opam/4.14.1/lib/stdio/stdio.cmxa /Users/mac1000/.opam/4.14.1/lib/uutf/uutf.cmxa /Users/mac1000/.opam/4.14.1/lib/result/result.cmxa /Users/mac1000/.opam/4.14.1/lib/ppx_deriving/runtime/ppx_deriving_runtime.cmxa lib/opine.cmxa /Users/mac1000/.opam/4.14.1/lib/cmdliner/cmdliner.cmxa /Users/mac1000/.opam/4.14.1/lib/ppx_here/runtime-lib/ppx_here_lib.cmxa examples/.main.eobjs/native/dune__exe.cmx examples/.main.eobjs/native/dune__exe__Stats.cmx examples/.main.eobjs/native/dune__exe__Classtransforms.cmx examples/.main.eobjs/native/dune__exe__Main.cmx)
# Undefined symbols for architecture x86_64:
#   "_libintl_bind_textdomain_codeset", referenced from:
#       __locale_bind_textdomain_codeset in libpython.a(_localemodule.o)
#   "_libintl_bindtextdomain", referenced from:
#       __locale_bindtextdomain in libpython.a(_localemodule.o)
#   "_libintl_dcgettext", referenced from:
#       __locale_dcgettext in libpython.a(_localemodule.o)
#   "_libintl_dgettext", referenced from:
#       __locale_dgettext in libpython.a(_localemodule.o)
#   "_libintl_gettext", referenced from:
#       __locale_gettext in libpython.a(_localemodule.o)
#   "_libintl_setlocale", referenced from:
#       __locale_setlocale in libpython.a(_localemodule.o)
#       __locale_localeconv in libpython.a(_localemodule.o)
#   "_libintl_textdomain", referenced from:
#       __locale_textdomain in libpython.a(_localemodule.o)
# ld: symbol(s) not found for architecture x86_64
# clang: error: linker command failed with exit code 1 (use -v to see invocation)
# File "caml_startup", line 1:
# Error: Error during linking (exit code 1)
```
Should we make it unavailable? You can do that adding `available: os != "macos"`

File: packages/opine/opine.0.1.3/opam [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+  "ppx_deriving"
+  "pyre-ast" {>= "0.1.6"}
+  "stdio"
+  "base" {>= "v0.15.0"}
+  "uutf"
+  "cmdliner" {>= "1.1.0"}
+  "sexplib"
+  "ppx_let"
+  "odoc" {with-doc}
+]
```
Comment (raphael-proust) [inline packages/opine/opine.0.1.3/opam:28]:
```suggestion
]
available: os != "macos"
```

As suggested by @mseri

**Comment by ArulselvanMadhavan** [comment]
Thank you @mseri @raphael-proust for the suggestion! I made the change directly to the opam. I was generating .opam file through dune. I couldn't find a way to include this restriction in dune, so I ended up editing the opam directly.

**Comment by ArulselvanMadhavan** [comment]
Tests failed "Unbound module Cmd". Will look into it

**Comment by ArulselvanMadhavan** [comment]
added (cmdliner (>= 1.1.1))

**Comment by raphael-proust** [comment]
Most of the errors are gone, but it seems the `macos` filter got removed in your force push. Presumably when you added the cmdliner dependency and regenerated the opam file.

**Comment by kit-ty-kate** [comment]
Thanks
